Filled out all recommended sections (#635)
zenml-io/zenml: docs/book/misc/best-practices.md

This ensures that loading the pipeline from elsewhere does not also run it.

### Never call the pipeline instance `pipeline` or a step instance `step`

Doing this will overwrite the imported `pipeline` and `step` decorators and lead
to failures at later stages if more steps and pipelines are decorated there.

```python
from zenml.pipelines import pipeline

@pipeline
def first_pipeline():
    ...

pipeline = first_pipeline(...)

@pipeline  # The code will fail here
def second_pipeline():
    ...
```

### Explicitly set `enable_cache` at the `@pipeline` level

> Caching is enabled by default for ZenML Pipelines. It is good to be explicit

Reviewer: ```suggestion
Caching is enabled by default for ZenML Pipelines. It is good to be explicit,
```

though so that it is clear when looking at the code if caching is enabled or

> disabled for a given pipeline.

Reviewer: ```suggestion
disabled for any given pipeline.
```

### Explicitly disable caching when loading data from filesystem or external APIs

ZenML inherently uses caching. However, this caching relies on changes of input
artifacts to invalidate the cache. In case a step has external data sources like
external APIs or filesystems, caching should be disabled explicitly for the 
step.

### Enable cache explicitly for steps that have a `context` argument, if they don't invalidate the caching behavior

Cache is implicitly disabled for steps that have a
[context]((../developer-guide/step-fixtures.md#using-the-stepcontext)) argument,
because it is assumed that you might use the step context to retrieve artifacts
from the artifact store that are unrelated to the current step. However, if that
is not the case, and your step logic doesn't invalidate the caching behavior, it
would be better to explicitly enable cache for your step

### Don't use the same metadata stores across multiple artifact stores

You might run into issues as the metadata store will point to artifacts in
inactive artifact stores.

### Use profiles to manage stacks

> Using profiles allows you to separate your zenml stack collections and work

Reviewer: ```suggestion
Using Profiles allows you to separate your ZenML stacks and work
```

Maybe don't introduce a new abstraction term that we don't in fact use anywhere else.

> locally within independant ZenML instances. See our

Reviewer: ```suggestion
locally within independent ZenML instances. See our
```

[docs](../collaborate/profiles.md) on profiles to learn more.

### Use unique pipeline names across projects, especially if used with the same metadata store

Pipeline names are their unique identifiers, so using the same name for
different pipelines will create a mixed history of runs between the two
pipelines.

### Check which integrations are required for registering a stack component

> You can do so by running `zenml flavor list` and installing the integration(s)
> if missing with `zenml integration install`.

Reviewer: ```suggestion
You can do so by running `zenml flavor list` and installing the missing integration(s) 
with `zenml integration install`.
```

### Initialize the ZenML repository in the root of the source code tree of a project, even if it's optional

> This will set the zenml project root for the project and create a local profile.

Reviewer: ```suggestion
This will set the ZenML project root for the project and create a local profile.
```

The advantage is that you create and maintain your active stack on a project
level.

### Include a `.dockerignore` in the ZenML repository to exclude files and folders from the container images built by ZenML for containerized environments

> Containerized Orchestrators and Step-Operators load your complete project files 

Reviewer: ```suggestion
Containerized Orchestrators and Step Operators load your complete project files 
```

> into a docker image for execution. To speed up the process and reduce docker 

Reviewer: ```suggestion
into a Docker image for execution. To speed up the process and reduce Docker 
```

> image sized exclude all unnecessary files (like data, virtual environments, 

Reviewer: ```suggestion
image sizes, exclude all unnecessary files (like data, virtual environments, 
```

When using the [post execution workflow](../developer-guide/post-execution-workflow.md)
it is tempting to access the pipeline views directly by their index, however,
the pipelines within your Repository are sorted by time of first run, so the 
pipeline at [-1] might not be the one you are expecting.

```python
from zenml.repository import Repository
from zenml.pipelines import pipeline

first_pipeline.run()
second_pipeline.run()
first_pipeline.run()

repo = Repository()
repo.get_pipelines()
>>> [PipelineView('first_pipeline'), PipelineView('second_pipeline')]

# This is recommended explicit way to retrieve your specific pipeline
repo.get_pipeline(pipeline_name="first_pipeline")
```

### Have your imports relative to your `.zen` directory OR have your imports relative to the root of your repository in cases when you don't have a `.zen` directory (=> which means to have the runner at the root of your repository)

ZenML uses the `.zen` repository root to resolve the class path of your 
functions and classes in a way that is portable across different types of 
environments such as containers. If a repository is not present, the location 

> of the main python module is used as an implicit repository root

Reviewer: ```suggestion
of the main Python module is used as an implicit repository root.
```
